File: molviewcpp/src/chem/resonance.cpp
#include "chem/resonance.h"

#include <cmath>
#include <cstdio>
#include <utility>
#include <vector>

#include "chem/smiles.h"

namespace chem {

void applyResonance(Molecule& mol) {
    const std::vector<std::vector<int>> nb = neighborLists(mol);

    for (std::size_t centerId = 0; centerId < mol.atoms.size(); centerId++) {
        const int center = static_cast<int>(centerId);
        if (nb[center].size() < 2) continue;

        // концевые соседи, сгруппированные по элементу (порядок групп — как у соседей)
        std::vector<std::pair<std::string, std::vector<int>>> groups;
        for (int nbId : nb[center]) {
            const Atom& neighbor = mol.atoms[nbId];
            if (neighbor.el == "H") continue;
            if (nb[nbId].size() != 1) continue;  // не концевой — не равноценен

            auto it = groups.begin();
            while (it != groups.end() && it->first != neighbor.el) ++it;
            if (it == groups.end()) {
                groups.emplace_back(neighbor.el, std::vector<int>{nbId});
            } else {
                it->second.push_back(nbId);
            }
        }

        for (const auto& group : groups) {
            const std::vector<int>& members = group.second;
            if (members.size() < 2) continue;

            std::vector<Bond*> bonds;
            for (int id : members) bonds.push_back(findBond(mol, center, id));

            bool anyAromatic = false;
            for (const Bond* b : bonds) if (b->aromatic) anyAromatic = true;
            if (anyAromatic) continue;

            bool sameOrder = true;
            bool sameCharge = true;
            for (std::size_t k = 0; k < members.size(); k++) {
                if (std::fabs(bonds[k]->order - bonds[0]->order) >= 1e-9) sameOrder = false;
                if (mol.atoms[members[k]].charge != mol.atoms[members[0]].charge) sameCharge = false;
            }
            if (sameOrder && sameCharge) continue;  // равноценны и без усреднения

            double sumOrder = 0;
            double sumCharge = 0;
            for (std::size_t k = 0; k < members.size(); k++) {
                sumOrder += bonds[k]->order;
                sumCharge += mol.atoms[members[k]].charge;
            }
            const double avgOrder = sumOrder / static_cast<double>(members.size());
            const double avgCharge = sumCharge / static_cast<double>(members.size());

            for (Bond* b : bonds) { b->order = avgOrder; b->delocalized = true; }
            for (int id : members) mol.atoms[id].charge = avgCharge;
        }
    }
}

std::string formatBondOrder(double order) {
    static const std::pair<double, const char*> table[] = {
        {1.0, "1"}, {4.0 / 3.0, "1⅓"}, {1.25, "1¼"}, {1.5, "1½"}, {5.0 / 3.0, "1⅔"},
        {2.0, "2"}, {2.5, "2½"}, {3.0, "3"},
    };
    for (const auto& row : table) {
        if (std::fabs(order - row.first) < 0.02) return row.second;
    }
    char buffer[32];
    std::snprintf(buffer, sizeof(buffer), "%.2f", order);
    std::string s(buffer);
    if (!s.empty() && s.back() == '0') s.pop_back();
    return s;
}

}  // namespace chem

File: molview/src/chem/resonance.ts
import type { Molecule } from './types';
import { neighborLists } from './smiles';

/**
 * ДЕЛОКАЛИЗАЦИЯ (РЕЗОНАНС)
 *
 * Формула Льюиса для нитрат-иона рисуется с одной двойной и двумя одинарными
 * связями, но на самом деле все три связи N-O одинаковы: длина 1,24 Å, порядок 1⅓.
 * Реальная частица — не одна из резонансных структур, а их наложение.
 *
 * Программа обнаруживает такие случаи сама: если центральный атом связан
 * с несколькими КОНЦЕВЫМИ атомами одного элемента, но связи получились разной
 * кратности, значит выбор «где рисовать двойную связь» произволен, и связи
 * надо усреднить. Заряд усредняется вместе с кратностью.
 *
 * Правило срабатывает для нитрат-, нитрит-, карбонат-, сульфат-, фосфат- и
 * ацетат-ионов, для нитрогруппы, для озона, для азотной кислоты — и НЕ
 * срабатывает для уксусной кислоты и сложных эфиров, где второй кислород
 * несёт водород или углеродный заместитель и потому неравноценен первому.
 */
export function applyResonance(mol: Molecule): void {
  const nb = neighborLists(mol);

  for (const center of mol.atoms) {
    if (nb[center.id].length < 2) continue;

    // концевые соседи, сгруппированные по элементу
    const groups = new Map<string, number[]>();
    for (const nbId of nb[center.id]) {
      const neighbor = mol.atoms[nbId];
      if (neighbor.el === 'H') continue;
      if (nb[nbId].length !== 1) continue; // не концевой — не равноценен
      if (!groups.has(neighbor.el)) groups.set(neighbor.el, []);
      groups.get(neighbor.el)!.push(nbId);
    }

    for (const [, members] of groups) {
      if (members.length < 2) continue;

      const bonds = members.map((id) => findBond(mol, center.id, id)!);
      if (bonds.some((b) => b.aromatic)) continue;

      const orders = bonds.map((b) => b.order);
      const charges = members.map((id) => mol.atoms[id].charge);
      const sameOrder = orders.every((o) => Math.abs(o - orders[0]) < 1e-9);
      const sameCharge = charges.every((c) => c === charges[0]);
      if (sameOrder && sameCharge) continue; // равноценны и без усреднения

      const avgOrder = orders.reduce((s, o) => s + o, 0) / orders.length;
      const avgCharge = charges.reduce((s, c) => s + c, 0) / charges.length;

      bonds.forEach((b) => { b.order = avgOrder; b.delocalized = true; });
      members.forEach((id) => { mol.atoms[id].charge = avgCharge; });
    }
  }
}

function findBond(mol: Molecule, a: number, b: number) {
  return mol.bonds.find((x) => (x.a === a && x.b === b) || (x.a === b && x.b === a));
}

/** Красивая запись дробного порядка связи: 1⅓, 1½, 1⅔. */
export function formatBondOrder(order: number): string {
  const table: [number, string][] = [
    [1, '1'], [4 / 3, '1⅓'], [1.25, '1¼'], [1.5, '1½'], [5 / 3, '1⅔'],
    [2, '2'], [2.5, '2½'], [3, '3'],
  ];
  for (const [value, label] of table) {
    if (Math.abs(order - value) < 0.02) return label;
  }
  return order.toFixed(2).replace(/0$/, '');
}
